File: sdk/nodejs/sampling.js
/**
 * SamplingClient — issue reverse `sampling/createMessage` JSON-RPC requests
 * to the host Agent (Anna).
 *
 * Why reverse RPC?
 *   - Plugins do NOT need their own LLM API key — billing/quotas/model
 *     routing are owned by the host (Anna).
 *   - Plugins describe a desired model via `modelPreferences` (MCP
 *     convention) and let the host pick a concrete model based on the
 *     user's saved preferences.
 *
 * Wire protocol (Executa v2):
 *   Plugin (us)                                Agent (host)
 *   ───────────────────────────────────────────────────────────────
 *   ← invoke(req_id=42, …)
 *   → sampling/createMessage(req_id=A)
 *   ← result | error                            (host replies)
 *   → invoke result(req_id=42)
 *
 * Threading model:
 *   - Construct one instance per process; share across tools.
 *   - Wire it up by feeding EVERY parsed JSON-RPC frame received on
 *     stdin to `client.dispatchResponse(msg)`. Frames that don't match
 *     a pending request are returned to your normal request handler.
 */

"use strict";

const crypto = require("node:crypto");

// ─── Constants — keep in sync with matrix/src/executa/protocol.py ─────

const PROTOCOL_VERSION_V1 = "1.1";
const PROTOCOL_VERSION_V2 = "2.0";

const METHOD_INITIALIZE = "initialize";
const METHOD_SHUTDOWN = "shutdown";
const METHOD_SAMPLING_CREATE_MESSAGE = "sampling/createMessage";

const SAMPLING_ERR_NOT_GRANTED = -32001;
const SAMPLING_ERR_QUOTA_EXCEEDED = -32002;
const SAMPLING_ERR_PROVIDER_ERROR = -32003;
const SAMPLING_ERR_INVALID_REQUEST = -32004;
const SAMPLING_ERR_TIMEOUT = -32005;
const SAMPLING_ERR_MAX_CALLS_EXCEEDED = -32006;
const SAMPLING_ERR_MAX_TOKENS_EXCEEDED = -32007;
const SAMPLING_ERR_NOT_NEGOTIATED = -32008;
const SAMPLING_ERR_USER_DENIED = -32009;

class SamplingError extends Error {
  constructor(code, message, data) {
    super(`[${code}] ${message}`);
    this.name = "SamplingError";
    this.code = code;
    this.data = data || {};
  }
}

class SamplingClient {
  /**
   * @param {object} [opts]
   * @param {(msg: object) => void} [opts.writeFrame] — defaults to writing
   *   a single newline-delimited JSON object to `process.stdout`.
   */
  constructor(opts = {}) {
    this._writeFrame =
      opts.writeFrame ||
      ((msg) => {
        process.stdout.write(JSON.stringify(msg) + "\n");
      });
    /** @type {Map<string, {resolve: Function, reject: Function, timer: NodeJS.Timeout}>} */
    this._pending = new Map();
    this._disabledReason = null;
  }

  /**
   * Mark sampling as unavailable (e.g. host did not negotiate v2).
   * @param {string} reason
   */
  disable(reason) {
    this._disabledReason = reason;
  }

  /**
   * Ask the host to run an LLM completion. Returns the host result object.
   *
   * @param {object} params
   * @param {Array<{role:"user"|"assistant"|"system", content:{type:"text", text:string}}>} params.messages
   * @param {number} params.maxTokens               — required, host enforces hard cap.
   * @param {string} [params.systemPrompt]
   * @param {number} [params.temperature]
   * @param {string[]} [params.stopSequences]
   * @param {object} [params.modelPreferences]      — MCP shape, omit to fall
   *                                                  back to the user's
   *                                                  preferred model.
   * @param {"none"} [params.includeContext]        — Phase 1 supports only "none".
   * @param {Object<string,string>} [params.metadata]
   * @param {number} [params.timeoutMs]             — default 90 000 ms.
   * @returns {Promise<object>}
   */
  createMessage(params) {
    if (this._disabledReason) {
      return Promise.reject(
        new SamplingError(SAMPLING_ERR_NOT_NEGOTIATED, this._disabledReason)
      );
    }
    const {
      messages,
      maxTokens,
      systemPrompt,
      temperature,
      stopSequences,
      modelPreferences,
      includeContext = "none",
      metadata,
      timeoutMs = 90_000,
    } = params || {};

    if (!Array.isArray(messages) || messages.length === 0) {
      return Promise.reject(new TypeError("messages must be a non-empty array"));
    }
    if (!Number.isInteger(maxTokens) || maxTokens <= 0) {
      return Promise.reject(new TypeError("maxTokens must be a positive integer"));
    }

    const reqId = crypto.randomUUID();
    const rpcParams = { messages, maxTokens, includeContext };
    if (systemPrompt != null) rpcParams.systemPrompt = systemPrompt;
    if (temperature != null) rpcParams.temperature = temperature;
    if (stopSequences) rpcParams.stopSequences = stopSequences;
    if (modelPreferences) rpcParams.modelPreferences = modelPreferences;
    if (metadata) rpcParams.metadata = metadata;

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        if (this._pending.delete(reqId)) {
          reject(
            new SamplingError(
              SAMPLING_ERR_TIMEOUT,
              `sampling/createMessage timed out after ${timeoutMs}ms`
            )
          );
        }
      }, timeoutMs);
      this._pending.set(reqId, { resolve, reject, timer });

      try {
        this._writeFrame({
          jsonrpc: "2.0",
          id: reqId,
          method: METHOD_SAMPLING_CREATE_MESSAGE,
          params: rpcParams,
        });
      } catch (err) {
        clearTimeout(timer);
        this._pending.delete(reqId);
        reject(err);
      }
    });
  }

  /**
   * Try to resolve the matching pending request from a parsed JSON-RPC
   * frame. Returns true if `msg` was a response we owned and it was
   * resolved/rejected accordingly.
   *
   * @param {object} msg — already JSON-parsed.
   * @returns {boolean}
   */
  dispatchResponse(msg) {
    if (!msg || typeof msg !== "object" || "method" in msg) return false;
    const id = msg.id;
    if (id == null) return false;
    const pending = this._pending.get(id);
    if (!pending) return false;
    this._pending.delete(id);
    clearTimeout(pending.timer);
    if (msg.error) {
      pending.reject(
        new SamplingError(
          Number(msg.error.code) || -32603,
          String(msg.error.message || "unknown error"),
          msg.error.data
        )
      );
    } else {
      pending.resolve(msg.result || {});
    }
    return true;
  }
}

module.exports = {
  SamplingClient,
  SamplingError,
  PROTOCOL_VERSION_V1,
  PROTOCOL_VERSION_V2,
  METHOD_INITIALIZE,
  METHOD_SHUTDOWN,
  METHOD_SAMPLING_CREATE_MESSAGE,
  SAMPLING_ERR_NOT_GRANTED,
  SAMPLING_ERR_QUOTA_EXCEEDED,
  SAMPLING_ERR_PROVIDER_ERROR,
  SAMPLING_ERR_INVALID_REQUEST,
  SAMPLING_ERR_TIMEOUT,
  SAMPLING_ERR_MAX_CALLS_EXCEEDED,
  SAMPLING_ERR_MAX_TOKENS_EXCEEDED,
  SAMPLING_ERR_NOT_NEGOTIATED,
  SAMPLING_ERR_USER_DENIED,
};

File: sdk/go/sampling/sampling.go
// Package sampling implements the Executa v2 reverse JSON-RPC
// `sampling/createMessage` request that lets a long-running plugin ask
// its host (Anna) to perform an LLM completion on its behalf.
//
// Why reverse RPC?
//   - Plugins do NOT need their own LLM API key — billing/quotas/model
//     routing are owned by the host.
//   - Plugins describe a desired model via ModelPreferences (MCP shape)
//     and let the host pick a concrete model based on the user's saved
//     preferences.
//
// Wire layout (Executa v2):
//
//	Plugin (us)                              Agent (host)
//	──────────────────────────────────────────────────────────
//	← invoke(req_id=42, …)
//	→ sampling/createMessage(req_id=A)
//	← result | error
//	→ invoke result(req_id=42)
//
// Threading model: construct one *Client per process; the same stdin
// reader feeds both agent-initiated requests AND host responses to our
// reverse calls. Use Client.DispatchResponse on every parsed frame; it
// returns true if the frame was a response we own.
package sampling

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"sync"
	"time"
)

// Constants — keep in sync with matrix/src/executa/protocol.py.
const (
	ProtocolVersionV1 = "1.1"
	ProtocolVersionV2 = "2.0"

	MethodInitialize            = "initialize"
	MethodShutdown              = "shutdown"
	MethodSamplingCreateMessage = "sampling/createMessage"

	ErrCodeNotGranted        = -32001
	ErrCodeQuotaExceeded     = -32002
	ErrCodeProviderError     = -32003
	ErrCodeInvalidRequest    = -32004
	ErrCodeTimeout           = -32005
	ErrCodeMaxCallsExceeded  = -32006
	ErrCodeMaxTokensExceeded = -32007
	ErrCodeNotNegotiated     = -32008
	ErrCodeUserDenied        = -32009
)

// SamplingError wraps a JSON-RPC error returned by the host.
type SamplingError struct {
	Code    int            `json:"code"`
	Message string         `json:"message"`
	Data    map[string]any `json:"data,omitempty"`
}

func (e *SamplingError) Error() string {
	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
}

// MessageContent is the {"type":"text","text":"..."} block.
type MessageContent struct {
	Type string `json:"type"`
	Text string `json:"text,omitempty"`
}

// Message is one MCP-shaped sampling message.
type Message struct {
	Role    string         `json:"role"` // "user" | "assistant" | "system"
	Content MessageContent `json:"content"`
}

// ModelHint matches MCP's modelPreferences.hints[*].
type ModelHint struct {
	Name string `json:"name,omitempty"`
}

// ModelPreferences maps to MCP modelPreferences.
type ModelPreferences struct {
	Hints                []ModelHint `json:"hints,omitempty"`
	CostPriority         *float64    `json:"costPriority,omitempty"`
	SpeedPriority        *float64    `json:"speedPriority,omitempty"`
	IntelligencePriority *float64    `json:"intelligencePriority,omitempty"`
}

// CreateMessageRequest mirrors the params of `sampling/createMessage`.
type CreateMessageRequest struct {
	Messages         []Message         `json:"messages"`
	MaxTokens        int               `json:"maxTokens"`
	SystemPrompt     string            `json:"systemPrompt,omitempty"`
	Temperature      *float64          `json:"temperature,omitempty"`
	StopSequences    []string          `json:"stopSequences,omitempty"`
	ModelPreferences *ModelPreferences `json:"modelPreferences,omitempty"`
	IncludeContext   string            `json:"includeContext,omitempty"` // "none" only in v1
	Metadata         map[string]string `json:"metadata,omitempty"`
}

// Usage mirrors the MCP usage block.
type Usage struct {
	InputTokens  int `json:"inputTokens"`
	OutputTokens int `json:"outputTokens"`
	TotalTokens  int `json:"totalTokens"`
}

// CreateMessageResult is the parsed result returned by the host.
type CreateMessageResult struct {
	Role       string         `json:"role"`
	Content    MessageContent `json:"content"`
	Model      string         `json:"model"`
	StopReason string         `json:"stopReason"`
	Usage      Usage          `json:"usage"`
	Meta       map[string]any `json:"_meta,omitempty"`
}

// FrameWriter writes one newline-delimited JSON-RPC frame to the host.
type FrameWriter func(msg map[string]any) error

// DefaultFrameWriter writes to os.Stdout under a process-wide mutex.
func DefaultFrameWriter() FrameWriter {
	var mu sync.Mutex
	return func(msg map[string]any) error {
		buf, err := json.Marshal(msg)
		if err != nil {
			return err
		}
		mu.Lock()
		defer mu.Unlock()
		if _, err := os.Stdout.Write(append(buf, '\n')); err != nil {
			return err
		}
		return nil
	}
}

type pending struct {
	ch chan json.RawMessage
}

// Client tracks outstanding reverse RPC requests and resolves them as
// responses arrive on stdin.
type Client struct {
	write           FrameWriter
	mu              sync.Mutex
	pending         map[string]*pending
	disabledReason  string
	defaultTimeoutS int
}

// New constructs a Client. Pass nil to use the default stdout writer.
func New(w FrameWriter) *Client {
	if w == nil {
		w = DefaultFrameWriter()
	}
	return &Client{
		write:           w,
		pending:         map[string]*pending{},
		defaultTimeoutS: 90,
	}
}

// Disable marks sampling as unavailable (e.g. host did not negotiate v2).
func (c *Client) Disable(reason string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.disabledReason = reason
}

// CreateMessage issues a `sampling/createMessage` request and waits for the
// host's response. Pass timeout = 0 to use the default (90s).
func (c *Client) CreateMessage(req CreateMessageRequest, timeout time.Duration) (*CreateMessageResult, error) {
	c.mu.Lock()
	if c.disabledReason != "" {
		reason := c.disabledReason
		c.mu.Unlock()
		return nil, &SamplingError{Code: ErrCodeNotNegotiated, Message: reason}
	}
	c.mu.Unlock()

	if len(req.Messages) == 0 {
		return nil, errors.New("messages must be non-empty")
	}
	if req.MaxTokens <= 0 {
		return nil, errors.New("maxTokens must be > 0")
	}
	if req.IncludeContext == "" {
		req.IncludeContext = "none"
	}
	if timeout <= 0 {
		timeout = time.Duration(c.defaultTimeoutS) * time.Second
	}

	id, err := newReqID()
	if err != nil {
		return nil, err
	}

	p := &pending{ch: make(chan json.RawMessage, 1)}
	c.mu.Lock()
	c.pending[id] = p
	c.mu.Unlock()

	envelope := map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"method":  MethodSamplingCreateMessage,
		"params":  req,
	}
	if err := c.write(envelope); err != nil {
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, err
	}

	select {
	case raw := <-p.ch:
		var resp struct {
			Result *CreateMessageResult `json:"result"`
			Error  *SamplingError       `json:"error"`
		}
		if err := json.Unmarshal(raw, &resp); err != nil {
			return nil, fmt.Errorf("decode sampling response: %w", err)
		}
		if resp.Error != nil {
			return nil, resp.Error
		}
		if resp.Result == nil {
			return nil, errors.New("empty sampling result")
		}
		return resp.Result, nil
	case <-time.After(timeout):
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, &SamplingError{
			Code:    ErrCodeTimeout,
			Message: fmt.Sprintf("sampling/createMessage timed out after %s", timeout),
		}
	}
}

// DispatchResponse resolves the matching pending request from a parsed
// JSON-RPC frame. Returns true if `frame` was a response we owned.
//
// Caller is responsible for parsing the frame once and inspecting whether
// it has a "method" field; if it does, dispatch to your normal handler
// instead of calling this.
func (c *Client) DispatchResponse(frame json.RawMessage) bool {
	var head struct {
		ID     any  `json:"id"`
		Method *any `json:"method"`
	}
	if err := json.Unmarshal(frame, &head); err != nil {
		return false
	}
	if head.Method != nil {
		return false
	}
	idStr, ok := head.ID.(string)
	if !ok {
		return false
	}
	c.mu.Lock()
	p := c.pending[idStr]
	if p != nil {
		delete(c.pending, idStr)
	}
	c.mu.Unlock()
	if p == nil {
		return false
	}
	select {
	case p.ch <- frame:
	default:
	}
	return true
}

func newReqID() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}
